lib/network/rest/server.js

```diff
@@ -0,0 +1,332 @@
+'use strict'; // eslint-disable-line
+
+const httpServer = require('../http/server');
+const Logger = require('werelogs').Logger;
+const url = require('url');
+const assert = require('assert');
+
+const constants = require('../../constants');
+const utils = require('./utils');
+const errors = require('../../errors');
+
+/**
+ * Parse the Range header into a two-integers array with possibly
+ * undefined values
+ *
+ * @param {String} rangeHeader - The 'Range' header value
+ * @return {(Number|undefined) []} An array of [left, right] integer
+ *   values provided in the range specification if present, or
+ *   undefined otherwise.
+ */
+function parseRangeSpec(rangeHeader) {
+    const rangeMatch = /^bytes=([0-9]+)?-([0-9]+)?$/.exec(rangeHeader);
+    // ignore header if syntax is invalid
+    if (rangeMatch) {
+        return rangeMatch.slice(1, 3).map(
+            str => (str === undefined ?
+                    undefined : Number.parseInt(str, 10)));
+    }
+    return undefined;
+}
+
+/**
+ * Convert a range specification as given by parseRangeSpec() into a
+ * fully specified absolute byte range
+ *
+ * @param {Number []} rangeSpec - Parsed range specification as returned
+ *   by parseRangeSpec()
+ * @param {Number} objectSize - Total byte size of the whole object to get
+ * @return {Number []} Fully specified byte range [start, end], as the
+ *   inclusive absolute byte range to request from the object.
+ */
+function getByteRange(rangeSpec, objectSize) {
+    if (objectSize === 0) {
+        return undefined;
+    }
+    const byteRange = [undefined, undefined];
+    if (rangeSpec[0] === undefined && rangeSpec[1] !== undefined) {
+        // suffix range
+        if (rangeSpec[1] >= objectSize) {
+            byteRange[0] = 0;
+        } else {
+            byteRange[0] = objectSize - rangeSpec[1];
+        }
+        byteRange[1] = objectSize - 1;
+    } else {
+        if (rangeSpec[0] > rangeSpec[1]) {
+            return undefined;
+        }
+        if (rangeSpec[0] >= objectSize) {
+            return undefined;
+        }
+        byteRange[0] = (rangeSpec[0] !== undefined ?
+                        rangeSpec[0] : 0);
+        byteRange[1] = (rangeSpec[1] < objectSize ?
+                        rangeSpec[1] : objectSize - 1);
+    }
+    return byteRange;
+}
+
+function setContentLength(response, contentLength) {
+    response.setHeader('Content-Length', contentLength.toString());
+}
+
+function setContentRange(response, byteRange, objectSize) {
+    const [start, end] = byteRange;
+    assert(start !== undefined && end !== undefined);
```

Comment:
I always found asserts a good tool to catch programming errors early. At least in C they are very much used in the ring code. Here, the previous `getByteRange` function ought to have returned a complete byte range, since that is its job, so we're checking this here.

Though if the practice is different in javascript for various reasons, I don't mind turning them to errors (here also, returning something means checking the return value, which may also add unnecessary complexity, if we can throw I find it better).

Please come back if you want it to be errors instead (though I will probably not be convinced still :))